cvmfs/network/download.cc

```diff
@@ -255,37 +255,49 @@ static size_t CallbackCurlData(void *ptr, size_t size, size_t nmemb,
 
   // LogCvmfs(kLogDownload, kLogDebug, "Data callback,  %d bytes", num_bytes);
 
-  if (num_bytes == 0)
+  // the check for kFailOk is to check when using the DataTube that there was
+  // not early some cancellation of the download due to error
+  // TODO(heretherebedragons) we might want to have this as an atomic variable?
+  if (num_bytes == 0 || info->error_code() != kFailOk) {
     return 0;
-
-  if (info->expected_hash()) {
-    shash::Update(reinterpret_cast<unsigned char *>(ptr),
-                  num_bytes, info->hash_context());
   }
 
-  if (info->compressed()) {
-    zlib::StreamStates retval =
-      zlib::DecompressZStream2Sink(ptr, static_cast<int64_t>(num_bytes),
-                                   info->GetZstreamPtr(), info->sink());
-    if (retval == zlib::kStreamDataError) {
-      LogCvmfs(kLogDownload, kLogSyslogErr,
-                                     "(id %" PRId64 ") failed to decompress %s",
-                                     info->id(), info->url()->c_str());
-      info->SetErrorCode(kFailBadData);
-      return 0;
-    } else if (retval == zlib::kStreamIOError) {
-      LogCvmfs(kLogDownload, kLogSyslogErr,
+  if (info->IsValidDataTube()) {
+    char *data = static_cast<char*>(malloc(num_bytes));
+    memcpy(data, ptr, num_bytes);
+    DataTubeElement *ele = new DataTubeElement(data, num_bytes, kActionData);
+    info->GetDataTubePtr()->EnqueueBack(ele);
+  } else {  // TODO(heretherebedragons) i think we need this here to support
+            // the non-multihreaded version?
```

Comment:
That is true. So this block should be factored out.

Reply:
in the final version i will do a nice cleanup putting it in a function